test/src/test/java/aima/test/unit/search/support/BiDirectionalAStartTest.java

```diff
@@ -0,0 +1,232 @@
+import java.util.function.ToDoubleFunction;
+
+import static aima.core.environment.map2d.SimplifiedRoadMapOfPartOfRomania.*;
+import static aima.core.environment.support.ProblemFactory.romaniaRoadMapProblem;
+
+/**
+ * @author wormi
+ */
+public class BiDirectionalAStartTest {
+
+  private List<GoAction> searchSolution(BidirectionalProblem<GoAction, InState> problem,
+                                        ToDoubleFunction<Node<GoAction, InState>> forwardHf,
+                                        ToDoubleFunction<Node<GoAction, InState>> backwardhf,
+                                        StateActionTimeLine stateActionTimeLine) {
+    BidirectionalSearch<GoAction, InState> search = new BidirectionalSearch<>(
+        BidirectionalSearch.Strategy.ASTAR, forwardHf, backwardhf);
+    if (stateActionTimeLine != null) {
+      search.setTimeLine(stateActionTimeLine);
+    }
+    try {
+      return search.findSolution(problem);
+    } catch (ExecutionException | InterruptedException e) {
+      throw new RuntimeException((e));
+    }
+  }
+
+  private List<GoAction> searchSolution(BidirectionalProblem<GoAction, InState> problem,
+                                        ToDoubleFunction<Node<GoAction, InState>> forwardHf,
+                                        ToDoubleFunction<Node<GoAction, InState>> backwardHf) {
+    return searchSolution(problem, forwardHf, backwardHf, null);
+  }
+
+  @Test
+  public void arad2arad() {
+    Assert.assertEquals(Arrays.asList((String) null),
+        searchSolution(romaniaRoadMapProblem(ARAD, ARAD), getHeuristicFunction(ARAD),
+            getHeuristicFunction(ARAD)));
+  }
+
+  private ToDoubleFunction<Node<GoAction, InState>> getHeuristicFunction(String... goals) {
+    return new StraightLineDistanceHeuristic(new SimplifiedRoadMapOfPartOfRomania(), goals);
+  }
+
+  @Test
+  public void arad2sibiu() {
+    final String start = ARAD;
+    final String goal = SIBIU;
+    Assert.assertEquals(
+        Arrays.asList(
+            new GoAction(SIBIU)),
+        searchSolution(romaniaRoadMapProblem(start, goal), getHeuristicFunction(goal),
+            getHeuristicFunction(start)));
+  }
+
+  @Test
+  public void arad2fagaras() {
+    final String start = ARAD;
+    final String goal = FAGARAS;
+    Assert.assertEquals(
+        Arrays.asList(
+            new GoAction(SIBIU),
+            new GoAction(FAGARAS)),
+        searchSolution(romaniaRoadMapProblem(start, goal), getHeuristicFunction(goal),
+            getHeuristicFunction(start)));
+  }
+
+  @Test
+  public void sibiu2bucharest() {
+    StateActionTimeLine timeLine = new StateActionTimeLine();
+    final String start = SIBIU;
+    final String goal = BUCHAREST;
+    final List<GoAction> actual = searchSolution(
+        romaniaRoadMapProblem(start, goal),
+        getHeuristicFunction(goal), getHeuristicFunction(start), timeLine);
+    final List<GoAction> expected;
+    if (actual.contains(new GoAction(FAGARAS))) {
```

Comment:
fixme?
A* should give an optimal solution. I can't figure out how to make parallel A* optimal. Is an optimal solution really required?